File: python/activities/research.py
"""The research_destination pipeline's LLM steps, each a Temporal Activity.

Native Claude — no agent framework. Two shapes:
  • structured steps (clarify / enrich / plan / write) force a JSON-schema
    response via `output_config.format` and parse it.
  • the search step hands Claude its built-in `web_search` server tool and
    collects the summary it writes. This is the demo's LIVE data: destination
    intelligence comes off the real web, while flights/hotels are seeded.

Retry philosophy matches activities/llm.py: the SDK's own retries are OFF
(max_retries=0) so Temporal owns ALL retries — every rate-limit or outage shows
as a retry in workflow history. Provider-rejected requests (bad auth/request)
are non-retryable; retrying can't fix those.

These honor the SAME per-conversation kill-switch as call_llm (control.llm_down),
so flipping the Demo-controls outage toggle mid-research makes the whole fan-out
ride out the outage and recover.
"""

import json

import anthropic
from temporalio import activity
from temporalio.exceptions import ApplicationError

import config
import prompts
from . import control
from models.types import (
    ClarifyResult,
    EnrichRequest,
    ReportData,
    SearchItem,
    SearchPlan,
    WriteRequest,
)

# The web_search server tool. We use the BASIC variant (web_search_20250305) on
# purpose: the newer web_search_20260209 does "dynamic filtering" by spinning up a
# code-execution sandbox per search, which adds many seconds — too slow to watch in
# a live demo. Basic returns results directly and is much snappier.
WEB_SEARCH_TOOL = {
    "type": "web_search_20250305",
    "name": "web_search",
    "max_uses": config.WEB_SEARCH_MAX_USES,
}

_FATAL = (
    anthropic.BadRequestError,
    anthropic.AuthenticationError,
    anthropic.PermissionDeniedError,
    anthropic.NotFoundError,
)


async def _guard() -> None:
    """Demo kill-switch (the Demo-controls 'API status' panel). While flipped to
    'down', raise a RETRYABLE error → Temporal retries with backoff until it's
    flipped back, and the retries show in the history."""
    if await control.llm_down():
        raise ApplicationError(
            "LLM provider is unavailable (simulated outage).", type="LLMProviderDown"
        )


def _client() -> anthropic.AsyncAnthropic:
    return anthropic.AsyncAnthropic(max_retries=0)


async def _structured(system: str, user: str, schema: dict,
                      max_tokens: int = 2048, effort: str = "low") -> dict:
    """One structured-output call: Claude must answer with JSON matching `schema`.
    Low effort by default keeps these planning/synthesis steps snappy for demos."""
    try:
        async with _client() as client:
            resp = await client.messages.create(
                model=config.ANTHROPIC_MODEL,
                max_tokens=max_tokens,
                system=system,
                messages=[{"role": "user", "content": user}],
                output_config={"effort": effort, "format": {"type": "json_schema", "schema": schema}},
            )
    except _FATAL as e:
        raise ApplicationError(
            f"LLM request rejected: {e}", type="LLMFatalError", non_retryable=True
        ) from e
    text = "".join(b.text for b in resp.content if b.type == "text")
    return json.loads(text)


async def _text(system: str, user: str, max_tokens: int = 1024, effort: str = "low") -> str:
    try:
        async with _client() as client:
            resp = await client.messages.create(
                model=config.ANTHROPIC_MODEL,
                max_tokens=max_tokens,
                system=system,
                messages=[{"role": "user", "content": user}],
                output_config={"effort": effort},
            )
    except _FATAL as e:
        raise ApplicationError(
            f"LLM request rejected: {e}", type="LLMFatalError", non_retryable=True
        ) from e
    return "".join(b.text for b in resp.content if b.type == "text").strip()


# ── clarify (human-in-the-loop) ──────────────────────────────────────────────
@activity.defn
async def plan_clarifications(query: str) -> ClarifyResult:
    await _guard()
    data = await _structured(prompts.CLARIFY_SYSTEM, query, prompts.CLARIFY_SCHEMA)
    return ClarifyResult(
        needs_clarification=bool(data.get("needs_clarification")) and bool(data.get("questions")),
        questions=data.get("questions", [])[:3],
    )


@activity.defn
async def enrich_query(req: EnrichRequest) -> str:
    await _guard()
    if not req.answers:
        return req.query
    qa = "\n".join(f"Q: {q}\nA: {a}" for q, a in req.answers.items())
    user = f"Original request:\n{req.query}\n\nClarifications:\n{qa}"
    return await _text(prompts.ENRICH_SYSTEM, user)


# ── plan ─────────────────────────────────────────────────────────────────────
@activity.defn
async def plan_searches(brief: str) -> SearchPlan:
    await _guard()
    system = prompts.plan_system(config.RESEARCH_SEARCHES)
    data = await _structured(system, f"Research brief:\n{brief}", prompts.PLAN_SCHEMA)
    searches = [SearchItem(**s) for s in data.get("searches", [])]
    return SearchPlan(searches=searches)


# ── search (native web_search tool) ──────────────────────────────────────────
@activity.defn
async def web_search(item: SearchItem) -> str:
    """One durable search. If the worker dies mid-fan-out, only the searches
    that hadn't completed re-run on restart — finished ones are already in the
    workflow history."""
    await _guard()
    user = (
        f"Search query: {item.query}\n"
        f"Why this matters: {item.reason}\n\n"
        "Search the web and summarize the most relevant findings."
    )
    messages: list[dict] = [{"role": "user", "content": user}]

    # Server-side tool loop: Claude may pause (pause_turn) if it hits the
    # per-turn tool-use cap — re-send to let it resume, with a hard cap.
    try:
        async with _client() as client:
            for _ in range(4):
                resp = await client.messages.create(
                    model=config.ANTHROPIC_MODEL,
                    max_tokens=1024,
                    system=prompts.SEARCH_SYSTEM,
                    tools=[WEB_SEARCH_TOOL],
                    output_config={"effort": "low"},
                    messages=messages,
                )
                if resp.stop_reason != "pause_turn":
                    break
                messages.append({"role": "assistant", "content": resp.content})
    except _FATAL as e:
        raise ApplicationError(
            f"LLM request rejected: {e}", type="LLMFatalError", non_retryable=True
        ) from e

    summary = "".join(b.text for b in resp.content if b.type == "text").strip()
    return f"### {item.query}\n{summary}" if summary else f"### {item.query}\n(no findings)"


# ── write ────────────────────────────────────────────────────────────────────
@activity.defn
async def write_report(req: WriteRequest) -> ReportData:
    await _guard()
    findings = "\n\n".join(req.findings)
    user = f"Research brief:\n{req.brief}\n\nFindings from web searches:\n\n{findings}"
    data = await _structured(prompts.WRITE_SYSTEM, user, prompts.WRITE_SCHEMA, max_tokens=1024, effort="low")
    return ReportData(
        short_summary=data.get("short_summary", ""),
        markdown_report=data.get("markdown_report", ""),
        follow_up_questions=data.get("follow_up_questions", [])[:3],
    )

File: typescript/src/activities/research.ts
// The research_destination pipeline's LLM steps, each a Temporal Activity.
// Ported from python/activities/research.py. Native Claude, no agent framework:
//   • structured steps (plan / write) force a JSON-schema response via
//     output_config.format and parse it.
//   • the search step hands Claude its built-in web_search server tool and
//     collects the summary it writes — the demo's LIVE data.
//
// Same retry philosophy + per-conversation kill-switch as llm.ts.

import Anthropic from '@anthropic-ai/sdk';
import { activityInfo } from '@temporalio/activity';
import { ApplicationFailure } from '@temporalio/common';

import * as config from '../config';
import * as prompts from '../prompts';
import * as control from './control';
import type { SearchItem, SearchPlan, WriteRequest, ReportData } from '../types';

// The web_search server tool. We use the BASIC variant (web_search_20250305) on
// purpose: the newer web_search_20260209 spins up a code-execution sandbox per
// search ("dynamic filtering"), which adds many seconds — too slow to watch in a
// live demo. Basic returns results directly and is much snappier.
const WEB_SEARCH_TOOL: Anthropic.WebSearchTool20250305 = {
  type: 'web_search_20250305',
  name: 'web_search',
  max_uses: config.WEB_SEARCH_MAX_USES,
};

const FATAL_STATUS = new Set([400, 401, 403, 404]);

type Effort = 'low' | 'medium' | 'high' | 'xhigh' | 'max';

function client(): Anthropic {
  return new Anthropic({ maxRetries: 0 });
}

function fatal(e: unknown): ApplicationFailure | null {
  if (e instanceof Anthropic.APIError && typeof e.status === 'number' && FATAL_STATUS.has(e.status)) {
    return ApplicationFailure.create({
      message: `LLM request rejected: ${e.message}`,
      type: 'LLMFatalError',
      nonRetryable: true,
    });
  }
  return null;
}

// Demo kill-switch: while flipped 'down', raise a RETRYABLE error → Temporal
// retries with backoff until it's flipped back.
async function guard(): Promise<void> {
  if (await control.llmDown()) {
    throw ApplicationFailure.create({
      message: 'LLM provider is unavailable (simulated outage).',
      type: 'LLMProviderDown',
    });
  }
}

// One structured-output call: Claude must answer with JSON matching `schema`.
async function structured(
  system: string,
  user: string,
  schema: Record<string, unknown>,
  maxTokens = 2048,
  effort: Effort = 'low'
): Promise<Record<string, unknown>> {
  let resp: Anthropic.Message;
  try {
    resp = await client().messages.create({
      model: config.ANTHROPIC_MODEL,
      max_tokens: maxTokens,
      system,
      messages: [{ role: 'user', content: user }],
      output_config: { effort, format: { type: 'json_schema', schema } },
    });
  } catch (e) {
    const f = fatal(e);
    if (f) throw f;
    throw e;
  }
  const text = resp.content
    .filter((b): b is Anthropic.TextBlock => b.type === 'text')
    .map((b) => b.text)
    .join('');
  return JSON.parse(text);
}

// ── plan ─────────────────────────────────────────────────────────────────────
export async function planSearches(brief: string): Promise<SearchPlan> {
  await guard();
  const system = prompts.planSystem(config.RESEARCH_SEARCHES);
  const data = await structured(system, `Research brief:\n${brief}`, prompts.PLAN_SCHEMA);
  const raw = (data.searches as SearchItem[] | undefined) ?? [];
  const searches = raw.map((s) => ({ query: s.query, reason: s.reason }));
  return { searches };
}

// ── search (native web_search tool) ──────────────────────────────────────────
// One durable search. If the worker dies mid-fan-out, only searches that hadn't
// completed re-run on restart — finished ones are already in the history.
export async function webSearch(item: SearchItem): Promise<string> {
  await guard();

  // Demo chaos: sometimes a search just flakes. Raise a RETRYABLE error so
  // Temporal retries this single activity (with backoff) while its siblings keep
  // going. Only the first attempt rolls the dice; retries always proceed.
  if (activityInfo().attempt === 1 && Math.random() < config.WEB_SEARCH_FAIL_RATE) {
    throw ApplicationFailure.create({
      message: `web_search transient failure (simulated) for ${JSON.stringify(item.query)}`,
      type: 'WebSearchFlaky',
    });
  }

  const user =
    `Search query: ${item.query}\n` +
    `Why this matters: ${item.reason}\n\n` +
    'Search the web and summarize the most relevant findings.';
  const messages: Anthropic.MessageParam[] = [{ role: 'user', content: user }];

  const c = client();
  let resp!: Anthropic.Message;
  try {
    // Server-side tool loop: Claude may pause (pause_turn) if it hits the
    // per-turn tool-use cap — re-send to let it resume, with a hard cap.
    for (let i = 0; i < 4; i++) {
      resp = await c.messages.create({
        model: config.ANTHROPIC_MODEL,
        max_tokens: 1024,
        system: prompts.SEARCH_SYSTEM,
        tools: [WEB_SEARCH_TOOL],
        output_config: { effort: 'low' },
        messages,
      });
      if (resp.stop_reason !== 'pause_turn') break;
      messages.push({ role: 'assistant', content: resp.content as unknown as Anthropic.ContentBlockParam[] });
    }
  } catch (e) {
    const f = fatal(e);
    if (f) throw f;
    throw e;
  }

  const summary = resp.content
    .filter((b): b is Anthropic.TextBlock => b.type === 'text')
    .map((b) => b.text)
    .join('')
    .trim();
  return summary ? `### ${item.query}\n${summary}` : `### ${item.query}\n(no findings)`;
}

// ── write ────────────────────────────────────────────────────────────────────
export async function writeReport(req: WriteRequest): Promise<ReportData> {
  await guard();
  const findings = req.findings.join('\n\n');
  const user = `Research brief:\n${req.brief}\n\nFindings from web searches:\n\n${findings}`;
  const data = await structured(prompts.WRITE_SYSTEM, user, prompts.WRITE_SCHEMA, 2500, 'low');
  return {
    short_summary: (data.short_summary as string) ?? '',
    markdown_report: (data.markdown_report as string) ?? '',
  };
}
